src/main/java/io/cryostat/agent/Agent.java

```diff
@@ -55,71 +59,56 @@ public class Agent {
-                            evt -> {
-                                switch (evt.state) {
-                                    case REGISTERED:
-                                        client.harvester().start();
-                                        break;
-                                    case UNREGISTERED:
-                                        client.harvester().stop();
-                                        break;
-                                    case REFRESHED:
-                                        break;
-                                    default:
-                                        log.error("Unknown registration state: {}", evt.state);
-                                        break;
-                                }
-                            });
-            client.registration().start();
-            client.webServer().start();
+            final Client client = DaggerAgent_Client.builder().build();
+            Registration registration = client.registration();
+            Harvester harvester = client.harvester();
+            WebServer webServer = client.webServer();
+            ExecutorService executor = client.executor();
+
+            agentExitHandler = installSignalHandlers(registration, harvester, webServer, executor);
+
+            registration.addRegistrationListener(
+                    evt -> {
+                        switch (evt.state) {
+                            case REGISTERED:
+                                client.harvester().start();
+                                break;
+                            case UNREGISTERED:
+                                client.harvester().stop();
+                                break;
+                            case REFRESHED:
+                                break;
+                            default:
+                                log.error("Unknown registration state: {}", evt.state);
+                                break;
+                        }
+                    });
+            registration.start();
+            webServer.start();
+            log.info("Startup complete");
         } catch (Exception e) {
             log.error(Agent.class.getSimpleName() + " startup failure", e);
-            return;
+            if (agentExitHandler != null) {
+                agentExitHandler.reset();
+            }
         }
-        log.info("Startup complete");
+    }
+
+    private static AgentExitHandler installSignalHandlers(
+            final Registration registration,
+            final Harvester harvester,
+            final WebServer webServer,
+            final ExecutorService executor) {
+        AgentExitHandler agentExitHandler =
+                new AgentExitHandler(registration, harvester, webServer, executor);
+        for (String s : List.of("INT", "TERM", "QUIT")) {
+            Signal signal = new Signal(s);
+            SignalHandler oldHandler = Signal.handle(signal, agentExitHandler);
```

Comment:
What exactly is the purpose of these couple lines? Seems like a lot of passing signal handlers back and forth.

Reply:
Yea, it is. The Signal API only allows for installing a new handler, and the return value is the old handler - it isn't adding a new one to a collection of listeners, it's replacing it. But the Agent should not be completely changing the way that the application it's attached to handles signals, it should only be catching the signals so that it can perform its own cleanup, and then forward those signals on to whatever other signal handler may have been there first, including the default one that just initiates a standard JVM shutdown.

`Runtime.addShutdownHook` looks like it should be the same job for us here, but in practice I found that at least some web application frameworks handle signals and cause a shutdown that doesn't give the agent's hook enough time to execute before the whole application is shut down, so the exit upload we want to perform gets interrupted every time. Going a layer lower in the stack and directly intercepting signals means the agent is able to perform blocking actions that prevent shutdown (which was the source of the original bug here). There is still the onus to properly pass the signal on and to complete our shutdown tasks in a reasonable timeframe, lest we get `SIGKILL`ed.

Reply:
Wow very interesting! Lot's of information I did not know, thanks for the writeup.